### PR TITLE
s390x: Small refactoring of ABI code

### PR DESCRIPTION
**Pull request by uweigand**
Remove a bit of duplicated code and eliminate some
unnecessary allocations.

FYI @cfallin

### CODE REVIEW

**Review by jameysharp — APPROVED** [review]
This is just fusing loops, plus reusing an existing filter/sort in `get_clobbered_callee_saves` instead of duplicating it in a `get_regs_saved_in_prologue` function that's only used once. That sounds good to me! As far as I can tell this patch should not change what code is emitted.

File: cranelift/codegen/src/isa/s390x/abi.rs [hunk 1]
```diff
@@ -509,28 +509,26 @@ impl ABIMachineSpec for S390xMachineDeps {
...
 
-        for &reg in clobbered_callee_saves.iter() {
-            match reg.to_reg().class() {
-                RegClass::Int => clobbered_gpr.push(reg),
-                RegClass::Float => clobbered_fpr.push(reg),
-            }
-        }
+        // Collect clobbered registers.
+        let mut clobbered_fpr = vec![];
+        let mut first_clobbered_gpr = 16;
```
Comment (cfallin) [inline cranelift/codegen/src/isa/s390x/abi.rs:515]: Could we add a comment here describing the invariant? I.e. that registers `first_clobbered_gpr..16` are clobbered (a contiguous range)?

(I realize the variable and logic are pre-existing but it's good to clarify while we're passing through :-) )

File: cranelift/codegen/src/isa/s390x/abi.rs [hunk 2]
```diff
@@ -607,28 +605,35 @@ impl ABIMachineSpec for S390xMachineDeps {
...
         if outgoing_args_size > 0 {
-            clobbered_gpr.push(Writable::from_reg(RealReg::from(gpr_preg(14))));
+            first_clobbered_gpr = 14;
         }
-        let mut first_clobbered_gpr = 16;
-        for reg in clobbered_gpr {
-            let enc = reg.to_reg().hw_enc();
-            if enc < first_clobbered_gpr {
-                first_clobbered_gpr = enc;
+        for &reg in clobbered_callee_saves.iter() {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/s390x/abi.rs:628]: Could we factor out this loop and link-register conditional (from here and `gen_clobber_save` above) so we have `let (clobbered_fpr, first_clobbered_gpr) = Self::get_clobbered_fpr_gpr(&clobbered_callee_saves[..], outgoing_args_size)` ?

**Review by cfallin — APPROVED** [review]
Thanks! Just a few suggestions below as long as we're refactoring / cleaning up.

**Comment by uweigand** [comment]
Like this?

**Pull request merged**